fedimint-api/src/db/mod.rs

```diff
@@ -79,14 +84,76 @@ impl Database {
     pub fn new(db: impl IDatabase + 'static, module_decoders: ModuleDecoderRegistry) -> Self {
         let inner = DatabaseInner {
             db,
+            notifications: Notifications::new(),
             module_decoders,
         };
 
         Self(Arc::new(inner))
     }
 
     pub async fn begin_transaction(&self) -> DatabaseTransaction {
-        DatabaseTransaction::new(self.0.db.begin_transaction().await, &self.0.module_decoders)
+        DatabaseTransaction::new(
+            self.0.db.begin_transaction().await,
+            &self.0.module_decoders,
+            &self.0.notifications,
+        )
+    }
+
+    /// Waits for key to be notified.
+    ///
+    /// Calls the `f` on every notifications. Return the value when `f` returns a `Some(T)`
+    pub async fn wait_key_check<K, T>(
+        &self,
+        key: &K,
+        f: impl Fn(Option<K::Value>) -> Option<T>,
+    ) -> Result<T>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        // not compile time yet :/
+        assert!(
+            <K as DatabaseKey>::NOTIFY,
+            "wait_key called with key({}) without NOTIFY flag",
+            std::any::type_name::<K>()
+        );
+        let key_bytes = key.to_bytes();
+        loop {
+            // register for notification
+            let notify = self.0.notifications.wait(&key_bytes);
```

Comment:
`.wait()` is now misleading. "register", "listen", etc.?

Reply:
renamed to register